p4-16/psa/PSA.mdk

```diff
@@ -608,3 +608,209 @@ matches the in parameter of the subsequent block.
+- 1M entries in a table (1M=2^20^).
+- 64 Mbits of table size in bits, which is the number of entries times
+  the size of one entry in bits.  The size of one entry is the key
+  width, plus the maximum total bits of action parameters, maximized
+  across all actions of the table.
+
+Minimum requirements for assignments, and arithmetic and Boolean
+operations.  These are total across those done within table actions,
+and those done outside of table actions (i.e. directly inside of
+control blocks, or within actions that are invoked directly from
+control blocks without applying a table).
+
+- 100 assignments executed per packet on ingress.  The same limit
+  applies independently for egress.
+- 50 arithmetic/Boolean operations on operands up to 32 bits in width,
+  per packet, on ingress.  The same limit applies independently for
+  egress.
+- 5 arithmetic/Boolean operations on operands up to 128 bits in width,
+  per packet, on ingress.  The same limit applies independently for
+  egress.
+
+Minimum requirements for PSA externs:
+
+- 4 `PacketReplicationEngine` method calls per packet (TBD if there
+  should be separate limits for each method call).
+- At least 64 bytes of user-defined metadata per packet can be carried
+  from the `Ingress` to `Egress` control block through the packet
+  buffer.
+- 512 copies that can be made of a packet by a
+  `PacketReplicationEngine` multicast group.
+- 256 multicast groups.
+- 4 `BufferingQueueingEngine` method calls per packet (TBD if there
+  should be separate limits for each method call).
+- 4 `Hash` `getHash` method calls per packet, with 160 total bytes of
+  input data.
+- 4 `Checksum` `clear` and `get` method calls per packet, with 160
+  total bytes of input data as arguments to `update` and `remove`
+  method calls.
+- 6 counter updates per packet (on different counters, which can be
+  any mix of `Counter` or `DirectCounter`).
+- 2 meter updates per packet (on different meters, which can be any
+  mix of `Meter` or `DirectMeter`).
+- 2 `Register` updates per packet (on different registers).  A
+  register `read`, followed by register `write`, counts as 1 update.
+- 2 `Random` `read` calls per packet.
+- TBD `ActionProfile` restrictions.
+- TBD `ActionSelector` restrictions.
+- TBD digest creations per packet.
+
+A PSA compliant implementation need not, but is allowed to, reject a
+P4 program if it does any of the following:
+
+- Applies the same table more than 1 time per packet.
+- Calls `getHash` on the same `Hash` instance more than 1 time per
+  packet.  (TBD: Is this a useful restriction?)
+- Calls `clear` or `get` on the same `Checksum` instance more than 1
+  time per packet, or calls `update` and `remove` more than a total of
+  6 times per packet.  (TBD: Is this a useful restriction?)
+- Updates the same `Counter`, `DirectCounter`, `Meter`, `DirectMeter`,
+  or `Register` instance more than 1 time per packet.
+- Uses the same `ActionProfile` for more than 1 table.
```

Comment:
@antoninbas This is a different question not directly related to action profiles, but somewhat related (the answer to this question might inform the answer to whether action profiles should be allowed in the PSA more than once per packet, is the reason I ask).

Are you (or is anyone) proposing that PSA should require a PSA compliant implementation to support multiple apply() operations for the _same table instance_, if one is in the 'then' branch of an 'if' statement, and the other is in the 'else' branch of the same 'if' statement?  For example, are you proposing that this program must be supported for all PSA compliant implementations, even if there are match dependencies between every pair of tables involved?
```
if (some_condition) {
    T1.apply();
    T2.apply();
    T3.apply();
    T4.apply();
    T5.apply();
} else {
    T5.apply();
    T4.apply();
    T3.apply();
    T2.apply();
    T1.apply();
}
```
If those are the only apply() statements for the tables T1 through T5 in the entire P4 program, then it is easily provable statically that they can all be done at most once per packet, but I would guess that some hardware platforms might have a hard time implementing it.  Certainly bmv2 today does not, although that may be for reasons unrelated to why some hardware platforms might have difficulty implementing it.